File: source-code/symbolic-AI/family.pl
%% family.pl - Prolog rules for family relationship reasoning
%%
%% Defines parent/2 and grandparent/2 as derived relations.
%% Facts (mother/2, father/2) are asserted at runtime from Python.
%%
%% Usage with family.py:
%%   The Python script loads this file, asserts mother/father facts,
%%   and queries grandparent relationships.

%% A parent is either a mother or a father.
parent(X, Y) :- mother(X, Y).
parent(X, Y) :- father(X, Y).

%% X is a grandparent of Z if X is a parent of Y and Y is a parent of Z.
grandparent(X, Z) :-
  parent(X, Y),
  parent(Y, Z).

File: source-code/symbolic-AI/family.py
# family.py - Prolog-based family relationship reasoning from Python
#
# Demonstrates loading Prolog rules (family.pl), asserting facts at runtime,
# and querying derived relationships. The Prolog rules define parent/2 and
# grandparent/2; we assert mother/father facts from Python and let Prolog's
# inference engine deduce grandparent relationships.
#
# Requirements: brew install swi-prolog && uv pip install swiplserver

from swiplserver import PrologMQI
from pprint import pprint

with PrologMQI() as mqi:
    with mqi.create_thread() as prolog_thread:
        # Load the family rules (parent/2 and grandparent/2)
        prolog_thread.query("[family].")

        # Assert initial family facts: Irene is Ken's mother, Ken is Ron's father
        print("Assert a few initial facts:")
        prolog_thread.query("assertz(mother(irene, ken)).")
        prolog_thread.query("assertz(father(ken, ron)).")

        # Query: who are grandparents of whom?
        # Prolog chains: mother(irene, ken) + father(ken, ron) => grandparent(irene, ron)
        result = prolog_thread.query("grandparent(A, B).")
        pprint(result)
        print(len(result))

        # Add another child for Ken and re-query
        print("Assert another test fact:")
        prolog_thread.query("assertz(father(ken, sam)).")
        result = prolog_thread.query("grandparent(A, B).")
        pprint(
            result
        )  # now includes both grandparent(irene, ron) and grandparent(irene, sam)
        print(len(result))
